### PR TITLE
unix-friendly output for 'ipfs dht' commands

### PR DESCRIPTION
**Pull request by noffle**
addresses https://github.com/ipfs/go-ipfs/issues/2494

### CODE REVIEW

**Comment by whyrusleeping** [comment]
Could you add some sharness tests using iptb for these now? Now that the output is relatively friendly, thats a much easier task.

**Comment by noffle** [comment]
cc @RichardLitt

**Comment by noffle** [comment]
@whyrusleeping I'm starting to look at this, but given my understanding of Kademlia, tests are going to be very fragile: they're highly timing dependent (read: `sleep` and crossing our fingers). Checking here to see if you know something I don't about making these deterministic.

**Comment by whyrusleeping** [comment]
they shouldnt be too fragile. I'd go for at least testing findprovs first. Add something on one node, run a findprovs from another and make sure that node shows up.

**Comment by noffle** [comment]
You were right: fragility wasn't a problem here. I guess we're good about setting up DHT population on init + local machine sockets -- awesome!

**Comment by noffle** [comment]
Finally passing! I don't think it liked me running `iptb init` with `--force`.

**Comment by RichardLitt** [comment]
!remindme when merged.

File: core/commands/dht.go [hunk 1]
```diff
@@ -317,17 +327,19 @@ var findPeerDhtCmd = &cmds.Command{
 
 var getValueDhtCmd = &cmds.Command{
 	Helptext: cmds.HelpText{
-		Tagline: "Run a 'GetValue' query through the DHT.",
+		Tagline: "Given a key, query the DHT for its best value.",
 		ShortDescription: `
-GetValue will return the value stored in the DHT at the given key.
+Outputs the best value for the given key.
+
+There may be several different values for a given key stored in the DHT, so this context 'best' means the record that is most desirable. There is no one metric for 'best': it depends entirely on the key type. For IPNS, 'best' is the record that is both valid and has the highest sequence number (freshest). Different key types can specify other 'best' rules.
```
Comment (RichardLitt) [inline core/commands/dht.go]: ...so _in_ this context...

File: core/commands/dht.go [hunk 2]
```diff
@@ -420,9 +432,17 @@ GetValue will return the value stored in the DHT at the given key.
...
 	Helptext: cmds.HelpText{
-		Tagline: "Run a 'PutValue' query through the DHT.",
+		Tagline: "Write a key/value pair to the DHT.",
 		ShortDescription: `
-PutValue will store the given key value pair in the DHT.
+Given a key of the form /foo/bar and a value of any form, this will write that value to the DHT with that key.
+
+Keys have two parts: a keytype (foo) and the key name (bar). IPNS uses the /ipns keytype, and expects the key name to be a Peer ID. IPNS entries are specifically formatted (protocol buffer).
+
+You may only use keytypes that are supported in your ipfs binary: currently this is only /ipns. (READ: unless you know what you're doing, you're better off using 'ipfs name publish' instead of this.)
```
Comment (RichardLitt) [inline core/commands/dht.go]: I'm not sure "Unless you know what you are doing" is a useful comment. I would reword this.

File: core/commands/dht.go (hunk 1 above)
Comment (noffle) [inline core/commands/dht.go]: Good catch!

File: core/commands/dht.go (hunk 2 above)
Comment (noffle) [inline core/commands/dht.go]: Reworded. Let me know how it sounds now.

**Comment by RichardLitt** [comment]
@noffle Changes look good, but there's an errant `'` and `)` now.

**Comment by noffle** [comment]
I see the extra `)` but no apostrophe issues. The one on `go-ipfs'` is intentional (possessive).

**Comment by RichardLitt** [comment]
I would avoid the possessive apostrophe here, a bit confusing. `go-ipfs DHT internals` works as a compound noun without it.

**Comment by noffle** [comment]
Removed.

File: test/sharness/t0170-dht.sh [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  ipfsi 1 dht put planet pluto >actual &&
+  PEERS=$(wc -l actual | cut -d '"'"' '"'"' -f 1) &&
+  [ -s actual ] ||
+	test_fsh cat actual
+'
+
+# ipfs dht findprovs <key>
+test_expect_success 'findprovs' '
+  ipfsi 4 dht findprovs planet >actual &&
+  PEERS=$(wc -l actual | cut -d '"'"' '"'"' -f 1) &&
```
Comment (whyrusleeping) [inline test/sharness/t0170-dht.sh]: should check the output here is what we expect.

File: test/sharness/t0170-dht.sh [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+test_expect_success 'peer ids' '
+  PEERID_0=$(ipfsi 0 id --format="<id>") &&
+  PEERID_2=$(ipfsi 2 id --format="<id>")
+'
+
+# ipfs dht findpeer <peerID>
+test_expect_success 'findpeer' '
+  ipfsi 1 dht findpeer $PEERID_0 >actual &&
+	egrep "/ip4/127.0.0.1/tcp/.*" actual >/dev/null &&
```
Comment (whyrusleeping) [inline test/sharness/t0170-dht.sh]: test that the address has the correct port, we can get this info from running `ipfs id -f "<addrs>"` on the peer in question.

File: test/sharness/t0170-dht.sh [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+# start iptb + wait for peering
+NUM_NODES=5
+test_expect_success 'init iptb' '
+  iptb init -n $NUM_NODES --bootstrap=none --port=0 &&
+  startup_cluster $NUM_NODES
+'
+
+test_expect_success 'peer ids' '
+  PEERID_0=$(ipfsi 0 id --format="<id>") &&
```
Comment (whyrusleeping) [inline test/sharness/t0170-dht.sh]: can use `iptb get id 0` here (I know its poorly documented, i'm going to work on that soon)

File: test/sharness/t0170-dht.sh (hunk 3 above)
Comment (noffle) [inline test/sharness/t0170-dht.sh]: What is the correct way to do this? I've noticed that the output of `put` often does not match the output of `findprovs`. There seems to always be *an* intersection, but some keys are missing or are different. I'm not sure why this is.

File: test/sharness/t0170-dht.sh (hunk 5 above)
Comment (noffle) [inline test/sharness/t0170-dht.sh]: Awesome -- thanks.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a93435084420de6048a0202121cd826b7ae8e836
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: a8589bee50a51892c211ef70811d08666e9af7e6
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

 * Commit: b4ccf4498450e4b638e6886a539a90478fb8e4ae
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by: .* <.*>**

We ask for a few features in the commit message for Open Source licensing hygiene and commit message clarity.
`git commit --amend` can often help you quickly improve the commit message.
[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) and [a script](https://github.com/ipfs/community/blob/master/dev/tools/hooks/setup_commit_msg_hook.sh) are available to help in the long run.
Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23).

---
This message was auto-generated by https://gitcop.com

File: test/sharness/t0170-dht.sh (hunk 3 above)
Comment (noffle) [inline test/sharness/t0170-dht.sh]: Done.

**Comment by whyrusleeping** [comment]
new tests seem to be failing

**Comment by noffle** [comment]
Got it. `foo >expected | sort` != `foo | sort >expected`. :grinning:

**Comment by RichardLitt** [comment]
Thanks for this!
